TST: Remove all permanent filter changes from tests

seberg opened the pull request:
On my system there are two warnin changes remaining after running
the test. The Special function add an always filter during
import and one downstream package does the same.
## 

Relies on copy pasting suppress_warnings from numpy, this is not yet merged so there might be API flux, which would have to be copied over if merged.

seberg commented on scipy/sparse/tests/test_base.py:
```diff
@@ -198,6 +211,9 @@ class _TestCommon:
 
+    # TODO: Why this decorator is necessary when there is a module level
+    #       suppression, is completely puzzeling to me.
+    @sup_sparse_efficiency2.decorate
```
I really do not understand this, this should happen inside the setup_module/teardown_module active suppress_warnings context. But I need an extra suppression here.

seberg commented:
This does not clean up all "ignore" filters, etc. Just global filter list manipulations, there are more then hundred of those left.

rgommers commented on scipy/_lib/_numpy_compat.py:
```diff
@@ -45,3 +49,236 @@ def _assert_warns(warning_class, func, *args, **kw):
         return result
+
+
+if NumpyVersion(np.__version__) >= '1.12.0.dev':
```
This will break for 1.12 dev versions without ``suppress_warnings``, so better make this a feature check (just try import it from ``numpy.testing``)

rgommers replied:
and of course that also provides robustness about ``numpy`` not actually including this - ``__numpy_ufunc__`` lesson:(

seberg replied:
Hehe, ok. Do you know if the test failure is real or not? First sight I was not sure.

rgommers replied:
TravisCI test failure is against numpy master, so will also fix that

ev-br commented on scipy/_lib/_numpy_compat.py:
```diff
@@ -45,3 +49,236 @@ def _assert_warns(warning_class, func, *args, **kw):
+            self._orig_show(message, category, filename, lineno, *args, **kwargs)
+
+        def decorate(self, func):
+            """
```
Minor quibble (feel free to ignore): is there a reason to have a name for this instead of a `__call__` method?

seberg replied:
Nope, can't think of another useful thing for call, so will change
(first on numpy). Quabbles are good, I already added a specific log
also `log = sup.record(UserWarning)` and not just the global one. If we
add such a monster, might as well get it right :).

seberg replied:
I am also not too happy with the name, since filter/suppression is now also recording. In some sense it is just catch_warnings pimped up ;).

seberg commented:
I remembered that this is still flying around. Is there still interest in doing something like this? I saw some discussion about changing the test suit, etc. too, which may also make this tricky (since my guess is to make it work nice, you really would have to put it into pytest or whatever tool you use then).

ev-br commented:
Does it really interact with the test runner (nose/pytest/whatever)?

seberg commented:
Hmm, yeah, you are right. Started to use this for testing as well, but
its mostly for suppressing which is probably a bit independend.

rgommers commented:
> I remembered that this is still flying around. Is there still interest in doing something like this? 

I think yes. Running into this once in a while, for example just encountered gh-5823 again.

seberg commented:
Hmm, I will try to have a look at it some time again, though IIRC this was just a start basically.

rgommers commented:
Maybe useful to merge this and open a TODO issue for the modules that still need work?

seberg commented:
Yes, likely that makes sense, IIRC, in this part I mostly tried to get rid of nonsense global filters, not yet so much all the problematic (on python 2.7 at least) "ignore" filters. The context manager should be updated with the now merged numpy version (not sure if that changes other code), and the merge conflicts may need a bit careful looking at. I had just looked at it a bit on the weekend, and while it seemed not to bad, I screwed up the first time. (plus then scipy did not compile, but probably I just should have removed the old compilation files, did not compile it in ages after all).

rgommers commented:
yeah, a ``git clean -xdf`` does wonders:)

seberg commented:
OK, I rebased this, added a couple of extra suppressions. The test suit (for me), now also runs with `scipy.test(raise_warnings="develop")` (meaning that warnings are set to errors on the outermost level. But it seems not too unlikely that other setups may experience other warnings.

Global filter manipulation did not seem to have worsened over time (actually probably got better).

There are something like 186 "ignore" filters in scipy, which might be annoying (and especially DeprecationWarnings can also cause cascade upon cleaning up). I am not inclined to starting to fix those, sorry.

Warnings during tests seem to be pretty clean, so possibly you could set scipy set up to also use "develop" mode in development versions. In numpy, I got some slow tests (I think I stole the design from from scipy), that check the code for "ignore" filters to avoid them.

seberg commented:
Oh, one last thing, if you go through this, use the `-b` flag for git diff (or some other tool) to hide the whitespace changes. That makes the test diffs sane to go through quickly.

seberg commented:
OK, cleaned up the BadCoefficient warning thing, down to 161 ;)

seberg commented:
Is the test error real? (if it is, I guess it is a PEP 8 error, which would probably be in the copy paste in numpy compat though).

seberg commented:
Sorry, for the noise, there was an error elsewhere and I missed the pointer in the test first. Lets see if it shows up for numpy compat (which I guess it won't).

seberg commented:
OK, this should be good now as a first step.

rgommers commented:
LGTM now, merging. Thanks Sebastian!

seberg commented:
@rgommers thanks. I created gh-7524 with some hints on how one would go about to remove the last filters. Dunno if this is what you thought of, so feel free to move/delete/change of course.
